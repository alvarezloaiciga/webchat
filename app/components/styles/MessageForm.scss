.MessageForm {
  width: 100%;
  flex: 0 0 auto;
  background: #fff;
  display: flex;
  flex-direction: column;

  .messageArea {
    display: flex;
    flex: 0 0 auto;

    textarea {
      padding: 12px 0 12px 12px;
      flex: 1 1 auto;
      width: 100%;
      border: none;
      resize: none;
      font-size: 14px;

      &:focus {
        outline: none;
      }
    }

    .sendBtn {
      padding: 12px;
      flex: 0 0 auto;
      border: none;
      font-size: 14px;
      background: #fff;
      font-weight: bold;
      transition: 0.15s ease-in-out color;

      &:not([disabled]) {
        color: #59ad5d;
      }
    }
  }

  .poke {
    margin-top: auto;
    background: white;
    font-size: 14px;
    border-bottom: 1px solid #eee;
    animation: fadeUp 1.0s 1;

    .pokeBody {
      padding: 7px 9px;

      .TypingIndicator {
        margin-left: 5px;
        vertical-align: text-bottom;
      }
    }
  }
}

// Total Width: 400px
// Total Height: 600px
// MessageForm: 120px
// Banner: 40px
// Transcript: 440px
.noFlexbox .MessageForm {
  height: 120px;
  min-height: 120px;
  max-height: 120px;

  .poke {
    min-height: 30px;
    max-height: 30px;
    height: 30px;
    display: block;
  }

  .messageArea {
    display: table;

    textarea {
      max-height: 60px;
      min-height: 60px;
      height: 60px;
      overflow: auto;
      width: 318px; // 318 + 12padding = 330
    }

    .sendBtn {
      display: table-cell;
      vertical-align: middle;
      width: 46px; // 46 + 24padding = 70
    }
  }
}
