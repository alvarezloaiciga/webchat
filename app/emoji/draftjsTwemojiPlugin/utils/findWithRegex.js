const findWithRegex = (regex, contentBlock, callback) => {
  // Get the text from the contentBlock
  const text = contentBlock.getText();
  let matchArr;
  let start; // eslint-disable-line
  // Go through all matches in the text and return the indices to the callback
  // eslint-disable-next-line no-cond-assign
  while ((matchArr = regex.exec(text)) !== null) {
    start = matchArr.index;
    callback(start, start + matchArr[0].length);
  }
};

export default findWithRegex;
