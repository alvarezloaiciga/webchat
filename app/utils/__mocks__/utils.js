export const getBrowserName = () => 'Chrome';
export const getMajor = () => 60;
export const getDeviceType = () => 'PC';
export const getOSName = () => 'Windows';
export const getUAInfo = () => ({
  browser: {
    name: 'Chrome',
    major: 60,
  },
  os: {
    name: 'Mac OS',
    version: 10,
  },
  device: {
    model: 'iPad',
    vendor: 'Apple',
    type: 'tablet',
  },
  cpu: {
    architecture: 'amd64',
  },
  engine: {
    name: 'WebKit',
    version: 537.36,
  },
  navigator:
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.81 Safari/537.36',
});
export const compatibilityMode = () => false;
export const isIE9 = () => false;
export const isIE10 = () => false;
export const nonCompatibleBrowser = () => false;
export const supportsFlexbox = () => true;
export const supportsSVG = () => true;
export const displayError = (error: string) => error;
export const isIEorSafari = () => false;
