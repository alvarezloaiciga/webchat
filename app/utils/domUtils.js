import ReactDOM, {render} from 'react-dom';
import React from 'react';
import Routes from 'Routes';
import {quiqContainerId} from 'Common/Constants';
import {AppContainer} from 'react-hot-loader';
import {IntlProvider} from 'react-intl';
import Redbox from 'redbox-react';
import {Provider} from 'react-redux';

export const appIsMounted = (): boolean => {
  const container = document.getElementById(quiqContainerId);
  if (!container) return false;
  return container.childElementCount > 0;
};

export const destructApp = () => {
  ReactDOM.unmountComponentAtNode(document.getElementById(quiqContainerId));
};

export const constructApp = store => {
  render(
    <Provider store={store}>
      <IntlProvider locale="en">
        <AppContainer errorReporter={Redbox}>
          <Routes />
        </AppContainer>
      </IntlProvider>
    </Provider>,
    document.getElementById(quiqContainerId),
  );

  if (module.hot) {
    module.hot.accept('Routes', () => {
      const NextApp = require('Routes').default; // eslint-disable-line global-require

      render(
        <Provider store={store}>
          <IntlProvider locale="en">
            <AppContainer errorReporter={Redbox}>
              <NextApp />
            </AppContainer>
          </IntlProvider>
        </Provider>,
        document.getElementById(quiqContainerId),
      );
    });
  }
};

export const removeElement = (e: HTMLElement) => {
  e.parentNode.removeChild(e);
};

export const reloadApp = store => {
  destructApp();
  constructApp(store);
};

export const setFavicon = (iconUrl: string, appleTouchUrl: ?string) => {
  // Remove all icons
  const icons = [
    ...Array.from(document.querySelectorAll("link[rel='icon']")),
    ...Array.from(document.querySelectorAll("link[rel='apple-touch-icon']")),
  ];
  icons.forEach(icon => {
    removeElement(icon);
  });

  // Build icon
  const icon = document.createElement('link');
  icon.rel = 'icon';
  icon.href = iconUrl;

  // Build apple touch icon
  const appleTouchIcon = document.createElement('link');
  appleTouchIcon.rel = 'apple-touch-icon';
  appleTouchIcon.href = appleTouchUrl || iconUrl;

  // NOTE HARD: The apple-touch-icon must come BEFORE the regular icon.
  // Chrome looks for the last meta tag with a rel containing "icon" to try and use as title bar icon.
  const head = document.getElementsByTagName('head')[0];
  head.appendChild(appleTouchIcon);
  head.appendChild(icon);
};

export const setApplicationName = (name: string) => {
  const appName = document.querySelector("meta[name='application-name']");
  if (appName) {
    removeElement(appName);
  }
  const newAppName = document.createElement('meta');
  newAppName.name = 'application-name';
  newAppName.content = name;
  document.getElementsByTagName('head')[0].appendChild(newAppName);
};

export const setBrowserThemeColor = (color: string) => {
  const colorTag = document.querySelector("meta[name='theme-color']");
  if (colorTag) {
    removeElement(colorTag);
  }
  const newColor = document.createElement('meta');
  newColor.name = 'theme-color';
  newColor.content = color;
  document.getElementsByTagName('head')[0].appendChild(newColor);
};
