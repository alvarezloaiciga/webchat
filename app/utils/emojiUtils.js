import {emojiIndex} from 'emoji-mart';
import emojiRegexFactory from 'emoji-regex';
import invertBy from 'lodash/invertBy';
import type {EmojiMetadata, Emoji} from 'Common/Types';
import quiqOptions from 'Common/QuiqOptions';

const emojiRegex = emojiRegexFactory();

// Build lookup table of {unifiedCode : emojiId}
const emojiIdByUnified = {};
Object.keys(emojiIndex.emojis).forEach(id => {
  const unified = emojiIndex.emojis[id].unified.toLowerCase();
  emojiIdByUnified[unified] = id;
});

export const isSingleEmoji = (text: string): boolean => {
  // We know a string contains only one emoji iff
  // 1) An emoji regex finds only one match in the string, and
  // 2) The number of characters in that emoji === the number of characters in the string
  const emojis = text.match(emojiRegex);

  if (!emojis)
    return false;

  const emojiCharCount = emojis.reduce((count, e) => count + e.length, 0);
  return emojis.length === 1 && text.length === emojiCharCount;
};

export const convertUnicodeToEmojiObject = (u: string): ?Emoji => {
  const emojiId = Object.keys(emojiIndex.emojis).find(k => emojiIndex.emojis[k].native === u);
  return emojiIndex.emojis[emojiId];
};

// This function is used for filtering emojis in our picker, as well as redacting emojis prior to sending a message.
export const emojiFilter = (e: EmojiMetadata | Emoji | string) => {
  let emojiId;
  // emoji-mart will call this function when user is searching, but it only passes unicode string:
  if (typeof e === 'string') {
    emojiId =  emojiIdByUnified[e.toLowerCase()];
  } else {
    // NOTE: In emoji-mart land, id === short_codes[0]
    emojiId = e.id || e.short_names[0];
  }

  // Don't include any emojis that don't have id or shortcode (we don't know what they are)
  if (!emojiId)
    return false;

  const {includeEmojis, excludeEmojis} = quiqOptions;
  // Prioritize include over exclude. Only one of the should be defined by customer, however.
  if (Array.isArray(includeEmojis)) {
    return includeEmojis.includes(emojiId);
  }
  else if (Array.isArray(excludeEmojis)) {
    return !excludeEmojis.includes(emojiId);
  }

  // If neither exclude nor include was defined, always include emoji.
  return true;
};
