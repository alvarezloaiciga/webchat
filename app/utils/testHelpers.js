// @flow
import type {Message, IntlObject, IntlMessage} from 'types';

export const getMockMessage = (
  // eslint-disable-line import/prefer-default-export
  id?: number = 0,
  overrides?: Message | {} = {},
) => {
  const message = {
    authorType: id % 2 === 0 ? 'Customer' : 'Agent',
    text: `text--${id}`,
    id: `id--${id}`,
    timestamp: id,
    type: 'Text',
  };

  return Object.assign(message, overrides);
};

/**
 * Stub out formatMessage so that it will return the defaultMessage with basic variable replacement
 */
const formatTestMessage = (msg: IntlMessage, values?: {[key: string]: string}) => {
  if (!values) {
    return msg.defaultMessage;
  }

  let returnValue = msg.defaultMessage;
  Object.keys(values).forEach(key => {
    if (values) {
      const value = values[key];
      returnValue = returnValue.replace(`{${key}}`, value);
    }
  });

  return returnValue;
};

export const TestIntlObject: IntlObject = {
  formatMessage: formatTestMessage,
  formatRelative: value => value.toString(),
  formatTime: timestamp => timestamp.toString(),
  formatDate: timestamp => timestamp.toString(),
};
