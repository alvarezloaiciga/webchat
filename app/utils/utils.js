// @flow
declare var Modernizr: Object;
import {UAParser} from 'ua-parser-js';
import type {BrowserNames, DeviceTypes, OSNames} from 'types';

const parser = new UAParser();

export const getBrowserName = (): BrowserNames => parser.getResult().browser.name;

export const getMajor = (): number => parseInt(parser.getResult().browser.major, 10);

export const getDeviceType = (): DeviceTypes => parser.getResult().device.type;

export const getOSName = (): OSNames => parser.getResult().os.name;

export const getUAInfo = () => parser.getResult();

export const compatibilityMode = () => {
  const compatList = [
    {
      name: 'IE',
      major: 10,
    },
  ];
  const {name, major} = parser.getResult().browser;
  return !!compatList.find(i => i.name === name && parseInt(major, 10) <= i.major);
};

export const isIE9 = () => getBrowserName() === 'IE' && getMajor() <= 9;
export const isIE10 = () => getBrowserName() === 'IE' && getMajor() === 10;

export const nonCompatibleBrowser = () => getBrowserName() === 'IE' && getMajor() < 9;
// Modernizr says IE10 doesn't support flexbox.
// It kind of does, at least for what we need it for... so go ahead and ignore Modernizr in that case
export const supportsFlexbox = () => isIE10() || Modernizr.flexbox;
