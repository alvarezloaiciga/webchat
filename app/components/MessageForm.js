// @flow
import React, {Component} from 'react';
import {getStyle} from 'Common/QuiqOptions';
import {intlMessageTypes, MenuItemKeys} from 'Common/Constants';
import {isIE10, isMobile} from 'Common/Utils';
import {getOrientation} from 'utils/mobileUtils';
import {
  setMessageFieldFocused,
  setInputtingEmail,
  setAgentsAvailable,
  setWindowScrollLockEnabled,
} from 'actions/chatActions';
import {connect} from 'react-redux';
import QuiqChatClient from 'quiq-chat';
import EmojiTextarea from 'EmojiTextArea';
import EmailInput from 'EmailInput';
import EmojiPicker from 'EmojiPicker';
import MenuButton from 'core-ui/components/MenuButton';
import Input from 'core-ui/components/Input';
import {
  getAgentHasRespondedToLatestConversation,
  getAgentEndedLatestConversation,
  getLastClosedConversationIsSpam,
  getInputtingEmail,
  getClosedConversationCount,
  getMuteSounds,
  getConfiguration,
  getMessage,
} from 'reducers/chat';
import Menu from 'core-ui/components/Menu';
import * as EmojiUtils from '../utils/emojiUtils';
import './styles/MessageForm.scss';
import type {ChatState, Emoji, ChatConfiguration} from 'Common/types';

export type MessageFormProps = {
  lastClosedConversationIsSpam: boolean,
  closedConversationCount: number,
  agentHasRespondedToLatestConversation: boolean,
  agentsAvailable?: boolean,
  agentEndedConversation: boolean,
  muteSounds: boolean,
  configuration: ChatConfiguration,
  openFileBrowser: () => void,
  setMessageFieldFocused: (messageFieldFocused: boolean) => void,
  inputtingEmail: boolean,
  setInputtingEmail: (inputtingEmail: boolean) => void,
  setAgentsAvailable: (available: boolean) => void,
  setWindowScrollLockEnabled: (enabled: boolean) => void,
};

type MessageFormState = {
  simpleMode: boolean,
  hasText: boolean,
  inputText: string, // Only used for IE10
  emojiPickerVisible: boolean,
  agentsAvailableOrSubscribed: boolean,
};

let updateTimer;
export class MessageForm extends Component<MessageFormProps, MessageFormState> {
  textArea: EmojiTextarea;
  props: MessageFormProps;
  state: MessageFormState = {
    simpleMode: isIE10() || isMobile(),
    hasText: false,
    emojiPickerVisible: false,
    inputText: '',
    agentsAvailableOrSubscribed: false,
  };
  checkAvailabilityTimer: number;

  checkAvailability = async () => {
    if (this.props.configuration.enforceAgentAvailability) {
      const {available} = await QuiqChatClient.checkForAgents();

      this.props.setAgentsAvailable(available);
      clearTimeout(this.checkAvailabilityTimer);

      // Continue checking only if the user is currently not subscribed
      if (!QuiqChatClient.isUserSubscribed()) {
        this.checkAvailabilityTimer = setTimeout(
          this.checkAvailability,
          this.props.configuration.agentsAvailableTimer,
        );
      }
    }
  };

  componentWillReceiveProps(nextProps: MessageFormProps) {
    this.setState({
      agentsAvailableOrSubscribed: nextProps.agentsAvailable || QuiqChatClient.isUserSubscribed(),
    });
  }

  componentWillUnmount() {
    clearTimeout(this.checkAvailabilityTimer);
  }

  componentDidMount() {
    if (!this.simpleMode) {
      setTimeout(() => {
        if (this.textArea) {
          this.textArea.focus();
        }
      }, 200);
    }

    if (!this.state.agentsAvailableOrSubscribed || this.props.agentEndedConversation) {
      this.checkAvailability();
    }

    this.props.setMessageFieldFocused(false);
  }

  componentWillMount() {
    this.setState({
      agentsAvailableOrSubscribed: this.props.agentsAvailable || QuiqChatClient.isUserSubscribed(),
    });
  }

  componentWillUpdate(nextProps: MessageFormProps) {
    if (!this.props.agentEndedConversation && nextProps.agentEndedConversation) {
      this.checkAvailability();
    }
  }

  getFilteredText = () => {
    const text = this.state.simpleMode ? this.state.inputText : this.textArea.getPlaintext().trim();

    // Filter emojis based on includeEmojis/excludeEmojis
    return EmojiUtils.filterEmojisFromText(text);
  };

  startTyping = () => {
    const text = this.getFilteredText();
    if (text) {
      QuiqChatClient.updateTypingIndicator(text, true);
    }
    updateTimer = undefined;
  };

  stopTyping = () => {
    QuiqChatClient.updateTypingIndicator(this.getFilteredText(), false);
  };

  startTypingTimers = () => {
    if (!updateTimer) {
      updateTimer = setTimeout(this.startTyping, 2000);
    }
  };

  resetTypingTimers = () => {
    clearTimeout(updateTimer);
    updateTimer = undefined;
    this.stopTyping();
  };

  handleTextChanged = (text: string) => {
    const typingTimers = text ? this.startTypingTimers : this.resetTypingTimers;
    this.setState({hasText: !!text, inputText: text}, typingTimers);
  };

  addMessage = () => {
    const text = this.getFilteredText();
    if (text) {
      QuiqChatClient.sendTextMessage(text);
    }

    if (this.state.simpleMode) {
      this.setState({inputText: '', hasText: false}, this.resetTypingTimers);
    } else {
      // Even if there was no text to send after filtering, we still clear the form and reset timers.
      // No need to explicitly call resetTimers() as setting text field to empty string will result in the same
      this.textArea.setText('');
    }

    if (isMobile() && this.state.simpleMode) {
      this.textArea.input.blur();
    }
  };

  toggleEmojiPicker = () => {
    this.setState(
      state => ({emojiPickerVisible: !state.emojiPickerVisible}),
      () => {
        if (!this.state.emojiPickerVisible) {
          this.textArea.focus();
        }
      },
    );
  };

  handleEmojiSelection = (emoji: Emoji) => {
    this.setState({emojiPickerVisible: false});
    this.addEmoji(emoji);
  };

  addEmoji = (emoji: Emoji) => {
    if (this.state.simpleMode) {
      this.setState(
        state => ({inputText: `${state.inputText} ${emoji.native} `}),
        this.textArea.focus,
      );
    } else {
      this.textArea.insertEmoji(emoji.native);
    }
  };

  toggleEmailInput = () => {
    this.props.setInputtingEmail(!this.props.inputtingEmail);
  };

  handleMessageFieldFocused = () => {
    this.props.setMessageFieldFocused(true);

    this.props.setWindowScrollLockEnabled(false);

    // On mobile devices, we need to scroll text area onto top of keyboard if in landscape
    if (this.state.simpleMode && this.textArea && getOrientation() === 'landscape') {
      window.scrollTo(0, this.textArea.input.offsetTop);
    }
  };

  handleMessageFieldLostFocus = () => {
    this.props.setMessageFieldFocused(false);

    this.props.setWindowScrollLockEnabled(true);
    window.scrollTo(0, 0);
  };

  handleIMEModeEntered = (pendingText: string) => {
    this.setState({
      inputText: `${this.textArea.getPlaintext()}${pendingText}`,
      simpleMode: true,
    });
  };

  toggleMuteSounds = () => {
    QuiqChatClient.setCustomPersistentData('muteSounds', !this.props.muteSounds);
  };

  renderMenu = () => {
    const {colors, styles, fontFamily} = this.props.configuration;

    // Ensure custom options come first.
    const options = this.props.configuration.menuOptions.customItems.map(o => ({
      onClick: () => window.open(o.url, '_blank'),
      label: o.label,
      title: o.title,
      id: o.id,
      icon: o.icon
        ? {
            name: o.icon,
            style: getStyle(Object.assign({}, styles.OptionsMenuLineItemIcon, o.iconStyle), {
              color: colors.menuText,
            }),
          }
        : undefined,
      style: getStyle(Object.assign({}, styles.OptionsMenuLineItem, o.itemStyle), {
        color: colors.menuText,
        fontFamily,
        zIndex: 1000,
      }),
    }));

    if (this.props.configuration.playSoundOnNewMessage) {
      options.push({
        onClick: this.toggleMuteSounds,
        label: this.props.muteSounds
          ? getMessage(intlMessageTypes.unmuteSounds)
          : getMessage(intlMessageTypes.muteSounds),
        title: this.props.muteSounds
          ? getMessage(intlMessageTypes.unmuteSoundsTooltip)
          : getMessage(intlMessageTypes.muteSoundsTooltip),
        id: MenuItemKeys.MUTE_SOUNDS,
        icon: {
          name: this.props.muteSounds ? 'volume-up' : 'volume-off',
          style: getStyle(styles.OptionsMenuLineItemIcon, {
            color: colors.menuText,
          }),
        },
        style: getStyle(styles.OptionsMenuLineItem, {
          color: colors.menuText,
          fontFamily,
          zIndex: 1000,
        }),
        disabled: false,
      });
    }

    if (this.props.configuration.enableChatEmailTranscript) {
      options.push({
        onClick: this.toggleEmailInput,
        label: getMessage(intlMessageTypes.emailTranscriptMenuMessage),
        title: getMessage(intlMessageTypes.emailTranscriptMenuTooltip),
        id: MenuItemKeys.EMAIL_TRANSCRIPT,
        icon: {
          name: 'envelope',
          style: getStyle(styles.OptionsMenuLineItemIcon, {
            color: colors.menuText,
          }),
        },
        style: getStyle(styles.OptionsMenuLineItem, {
          color: colors.menuText,
          fontFamily,
          zIndex: 1000,
        }),
        disabled:
          (this.props.closedConversationCount === 0 || this.props.lastClosedConversationIsSpam) &&
          !this.props.agentHasRespondedToLatestConversation,
      });
    }

    return (
      options.length > 0 && (
        <MenuButton
          buttonStyles={getStyle(
            {
              borderRight: '2px solid rgb(244, 244, 248)',
            },
            styles.OptionsMenuButton,
          )}
          iconStyles={getStyle(styles.OptionsMenuButtonIcon, {
            color: '#848484',
            fontSize: '19px',
            marginTop: '1px',
          })}
          title={getMessage(intlMessageTypes.optionsMenuTooltip)}
          menuPosition="top-right"
          offset={Object.assign(
            {
              horizontal: '-115px',
              vertical: '40px',
            },
            this.props.configuration.menuOptions.offset,
          )}
        >
          <Menu
            items={options}
            containerStyle={getStyle(styles.OptionsMenuContainer, {
              fontFamily,
              zIndex: 1000,
            })}
          />
        </MenuButton>
      )
    );
  };

  render() {
    const {colors, styles, fontFamily} = this.props.configuration;
    const allowConversationToStart =
      !this.props.configuration.enableManualConvoStart || !this.props.agentEndedConversation;
    const sendDisabled =
      !this.state.hasText || !this.state.agentsAvailableOrSubscribed || !allowConversationToStart;
    const emopjiPickerDisabled =
      !this.state.agentsAvailableOrSubscribed || !allowConversationToStart;
    const contentButtonsDisabled =
      !this.state.agentsAvailableOrSubscribed || !allowConversationToStart;
    const inputStyle = getStyle(styles.MessageFormInput, {fontFamily});
    const sendButtonStyle = getStyle(styles.MessageFormSend, {
      color: colors.primary,
      fontFamily,
    });
    const contentButtonStyle = getStyle(styles.ContentButtons, {
      color: '#848484',
      fontSize: '16px',
    });

    let messagePlaceholder = this.state.agentsAvailableOrSubscribed
      ? getMessage(intlMessageTypes.messageFieldPlaceholder)
      : getMessage(intlMessageTypes.agentsNotAvailableMessage);

    if (!allowConversationToStart) {
      messagePlaceholder = getMessage(intlMessageTypes.cannotStartNewConversationMessage);
    }

    return (
      <div className="MessageForm" style={getStyle(styles.MessageForm)}>
        {this.props.inputtingEmail && (
          <div className="messageArea">
            <EmailInput onSubmit={this.toggleEmailInput} onCancel={this.toggleEmailInput} />
          </div>
        )}

        {!this.props.inputtingEmail && (
          <div className="messageArea">
            {this.state.simpleMode ? (
              <Input
                ref={element => {
                  this.textArea = element;
                }}
                inputStyle={inputStyle}
                value={this.state.inputText}
                maxLength={1024}
                autoFocus={!isMobile()}
                onBlur={this.handleMessageFieldLostFocus}
                onFocus={this.handleMessageFieldFocused}
                disabled={!this.state.agentsAvailableOrSubscribed || !allowConversationToStart}
                onChange={(e: SyntheticInputEvent<*>) => this.handleTextChanged(e.target.value)}
                onSubmit={this.addMessage}
                placeholder={messagePlaceholder}
              />
            ) : (
              <EmojiTextarea
                ref={n => {
                  this.textArea = n && n.wrappedInstance;
                }}
                style={inputStyle}
                disabled={!this.state.agentsAvailableOrSubscribed || !allowConversationToStart}
                name="message"
                maxLength={1024}
                onChange={this.handleTextChanged}
                onReturn={this.addMessage}
                onBlur={this.handleMessageFieldLostFocus}
                onFocus={this.handleMessageFieldFocused}
                onIMEModeEntered={this.handleIMEModeEntered}
                placeholder={messagePlaceholder}
              />
            )}

            {this.props.configuration.enableChatFileAttachments && (
              <button
                className="messageFormBtn attachmentBtn"
                style={contentButtonStyle}
                disabled={contentButtonsDisabled}
                onClick={this.props.openFileBrowser}
                title={getMessage(intlMessageTypes.attachmentBtnTooltip)}
              >
                <i className="fa fa-paperclip" />
              </button>
            )}
            {this.props.configuration.enableEmojis &&
              EmojiUtils.emojisEnabledByCustomer() &&
              !isMobile() && (
                <button
                  className="messageFormBtn emojiBtn"
                  style={contentButtonStyle}
                  disabled={emopjiPickerDisabled}
                  onClick={this.toggleEmojiPicker}
                  title={getMessage(intlMessageTypes.emojiPickerTooltip)}
                >
                  <i className="fa fa-smile-o" />
                </button>
              )}
            {sendDisabled ? (
              this.renderMenu()
            ) : (
              <button
                className="messageFormBtn sendBtn"
                onClick={this.addMessage}
                disabled={sendDisabled}
                style={sendButtonStyle}
              >
                {getMessage(intlMessageTypes.sendButtonLabel)}
              </button>
            )}
            {this.props.configuration.enableEmojis &&
              EmojiUtils.emojisEnabledByCustomer() &&
              !isMobile() && (
                <EmojiPicker
                  visible={this.state.emojiPickerVisible}
                  addEmoji={this.handleEmojiSelection}
                  emojiFilter={EmojiUtils.emojiFilter}
                  onOutsideClick={this.toggleEmojiPicker}
                  ignoreOutsideClickOnSelectors={['.emojiBtn']}
                />
              )}
          </div>
        )}
      </div>
    );
  }
}

const mapDispatchToProps = {
  setMessageFieldFocused,
  setInputtingEmail,
  setAgentsAvailable,
  setWindowScrollLockEnabled,
};

export default connect(
  (state: ChatState) => ({
    agentsAvailable: state.agentsAvailable,
    muteSounds: getMuteSounds(state),
    configuration: getConfiguration(state),
    agentEndedConversation: getAgentEndedLatestConversation(state),
    lastClosedConversationIsSpam: getLastClosedConversationIsSpam(state),
    closedConversationCount: getClosedConversationCount(state),
    agentHasRespondedToLatestConversation: getAgentHasRespondedToLatestConversation(state),
    inputtingEmail: getInputtingEmail(state),
  }),
  mapDispatchToProps,
)(MessageForm);
