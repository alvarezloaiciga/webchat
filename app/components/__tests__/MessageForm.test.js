// @flow
jest.mock('quiq-chat');
jest.mock('Common/Utils');
jest.mock('Common/QuiqOptions');
jest.mock('components/EmojiTextArea');
jest.mock('services/Postmaster');

import React from 'react';
import {MessageForm} from '../MessageForm';
import {shallow} from 'enzyme';
import type {ShallowWrapper} from 'enzyme';
import {getMockMessage, getMockConfiguration} from 'utils/testHelpers';
import type {MessageFormProps} from '../MessageForm';
import {MenuItemKeys} from 'Common/Constants';
import QuiqChatClient from 'quiq-chat';

describe('MessageForm component', () => {
  let wrapper: ShallowWrapper;
  let testProps: MessageFormProps;
  let instance: any;
  let render: () => void;

  beforeEach(() => {
    QuiqChatClient.sendTextMessage = jest.fn();
    QuiqChatClient.updateMessagePreview = jest.fn();

    render = () => {
      testProps = {
        transcript: [getMockMessage(), getMockMessage(1)],
        agentEndedConversation: false,
        latestConversationIsSpam: false,
        platformEvents: [],
        inputtingEmail: false,
        openFileBrowser: jest.fn(),
        muteSounds: false,
        setMuteSounds: jest.fn(),
        messageFieldFocused: false,
        setMessageFieldFocused: jest.fn(),
        setInputtingEmail: jest.fn(),
        configuration: getMockConfiguration({
          enableChatEmailTranscript: true,
          enableChatFileAttachments: true,
          enableEmojis: true,
          playSoundOnNewMessage: true,
          flashNotificationOnNewMessage: true,
          supportedAttachmentTypes: ['image/png,image/jpeg'],
          menuOptions: {
            customItems: [
              {
                id: 'Spidey',
                url: 'Earl',
                label: 'Babel',
                title: 'Danny McBridal',
                icon: 'Python',
                itemStyle: {},
                iconStyle: {},
              },
            ],
            offest: null,
          },
        }),
        chatIsSpam: false,
      };
      wrapper = shallow(<MessageForm {...testProps} />);
      instance = wrapper.instance();
    };
  });

  describe('rendering', () => {
    beforeEach(() => {
      render();
    });

    it('renders', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('adding text', () => {
    it('handling textarea onchange', () => {
      instance.handleTextChanged('foo');
      expect(wrapper.state('hasText')).toBe(true);
    });
  });

  describe('emailTranscript', () => {
    const isEmailTranscriptDisabled = () =>
      wrapper
        .find('Menu')
        .prop('items')
        .find(i => i.id === MenuItemKeys.EMAIL_TRANSCRIPT).disabled;

    beforeEach(() => {
      render();
      wrapper.setProps({agentEndedConversation: true});
      expect(isEmailTranscriptDisabled()).toBe(false);
    });

    describe('current convo is spam', () => {
      beforeEach(() => {
        wrapper.setProps({latestConversationIsSpam: true});
      });
      it('disables emailTranscript', () => {
        wrapper.setProps({chatIsSpam: true});
        expect(isEmailTranscriptDisabled()).toBe(true);
      });
    });

    describe('entire transcript has no agent message', () => {
      beforeEach(() => {
        wrapper.setProps({transcript: [], chatIsSpam: false});
      });
      it('disables emailTranscript', () => {
        expect(isEmailTranscriptDisabled()).toBe(true);
      });
    });
  });
});
