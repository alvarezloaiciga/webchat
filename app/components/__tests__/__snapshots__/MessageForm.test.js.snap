// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageForm component agentTyping shows typing indicator 1`] = `
<div
  className="MessageForm"
  style={
    Object {
      "boxShadow": "0 1px 2px 0 rgba(0, 0, 0, 0.2)",
      "zIndex": 1,
    }
  }
>
  <div
    className="poke"
  >
    <div
      className="pokeBody"
    >
      <span
        style={
          Object {
            "fontFamily": "Lato, sans-serif",
          }
        }
      >
        Agent is typing
      </span>
      <TypingIndicator
        xScale={0.75}
        yScale={0.5}
      />
    </div>
  </div>
  <div
    className="messageArea"
  >
    <EmojiTextArea
      disabled={false}
      maxLength={1024}
      name="message"
      onChange={[Function]}
      onReturn={[Function]}
      placeholder="Send us a message..."
      style={
        Object {
          "color": "#68b588",
          "fontFamily": "Source Sans Pro",
        }
      }
    />
    <button
      className="messageFormBtn emojiBtn"
      disabled={false}
      onClick={[Function]}
    >
      <i
        className="fa fa-smile-o"
        title="Minimize window"
      />
    </button>
    <button
      className="messageFormBtn sendBtn"
      disabled={true}
      onClick={[Function]}
      style={
        Object {
          "color": "#68b588",
          "fontFamily": "Source Sans Pro",
          "fontSize": 16,
          "textTransform": "uppercase",
        }
      }
    >
      Send
    </button>
    <EmojiPicker
      addEmoji={[Function]}
      emojiFilter={[Function]}
      ignoreOutsideClickOnSelectors={
        Array [
          ".emojiBtn",
        ]
      }
      onOutsideClick={[Function]}
      visible={false}
    />
  </div>
</div>
`;

exports[`MessageForm component rendering renders 1`] = `
<div
  className="MessageForm"
  style={
    Object {
      "boxShadow": "0 1px 2px 0 rgba(0, 0, 0, 0.2)",
      "zIndex": 1,
    }
  }
>
  <div
    className="messageArea"
  >
    <EmojiTextArea
      disabled={false}
      maxLength={1024}
      name="message"
      onChange={[Function]}
      onReturn={[Function]}
      placeholder="Send us a message..."
      style={
        Object {
          "color": "#68b588",
          "fontFamily": "Source Sans Pro",
        }
      }
    />
    <button
      className="messageFormBtn emojiBtn"
      disabled={false}
      onClick={[Function]}
    >
      <i
        className="fa fa-smile-o"
        title="Minimize window"
      />
    </button>
    <button
      className="messageFormBtn sendBtn"
      disabled={true}
      onClick={[Function]}
      style={
        Object {
          "color": "#68b588",
          "fontFamily": "Source Sans Pro",
          "fontSize": 16,
          "textTransform": "uppercase",
        }
      }
    >
      Send
    </button>
    <EmojiPicker
      addEmoji={[Function]}
      emojiFilter={[Function]}
      ignoreOutsideClickOnSelectors={
        Array [
          ".emojiBtn",
        ]
      }
      onOutsideClick={[Function]}
      visible={false}
    />
  </div>
</div>
`;
